client/tracing/src/logging/directives.rs

```diff
@@ -0,0 +1,123 @@
+};
+
+// Handle to reload the tracing log filter
+static FILTER_RELOAD_HANDLE: OnceCell<Handle<EnvFilter, SCSubscriber>> = OnceCell::new();
+// Directives that are defaulted to when resetting the log filter
+static DEFAULT_DIRECTIVES: OnceCell<Mutex<Vec<String>>> = OnceCell::new();
+// Current state of log filter
+static CURRENT_DIRECTIVES: OnceCell<Mutex<Vec<String>>> = OnceCell::new();
+
+/// Add log filter directive(s) to the defaults
+///
+/// The syntax is identical to the CLI `<target>=<level>`:
+///
+/// `sync=debug,state=trace`
+pub(crate) fn add_default_directives(directives: &str) {
+	DEFAULT_DIRECTIVES
+		.get_or_init(|| Mutex::new(Vec::new()))
+		.lock()
+		.push(directives.to_owned());
+	add_directives(directives);
+}
+
+/// Add directives to current directives
+pub fn add_directives(directives: &str) {
+	CURRENT_DIRECTIVES
+		.get_or_init(|| Mutex::new(Vec::new()))
+		.lock()
+		.push(directives.to_owned());
+}
+
+/// Parse `Directive` and add to default directives if successful.
+///
+/// Ensures the supplied directive will be restored when resetting the log filter.
+pub(crate) fn parse_default_directive(directive: &str) -> super::Result<Directive> {
+	let dir = directive.parse()?;
+	add_default_directives(directive);
+	Ok(dir)
+}
+
+/// Reload the logging filter with the supplied directives added to the existing directives
+pub fn reload_filter() -> Result<(), String> {
+	let mut env_filter = EnvFilter::default();
+	if let Some(current_directives) = CURRENT_DIRECTIVES.get() {
+		// Use join and then split in case any directives added together
+		for directive in current_directives
+			.lock()
+			.join(",")
+			.split(',')
+			.map(|d| d.parse())
+		{
+			match directive {
+				Ok(dir) => env_filter = env_filter.add_directive(dir),
+				Err(invalid_directive) => {
+					log::warn!(
+						target: "tracing",
+						"Unable to parse directive while setting log filter: {:?}",
+						invalid_directive,
+					);
+				}
+			}
+		}
+	}
+	env_filter = env_filter.add_directive(
+		"sc_tracing=trace"
+			.parse()
+			.expect("provided directive is valid"),
+	);
+	log::debug!(target: "tracing", "Reloading log filter with: {}", env_filter);
+	FILTER_RELOAD_HANDLE
+		.get()
+		.ok_or("No reload handle present".to_string())?
+		.reload(env_filter)
+		.map_err(|e| format!("{}", e))
+}
+
+/// Resets the log filter back to the original state when the node was started.
+///
+/// Includes substrate defaults and CLI supplied directives.
+pub fn reset_log_filter() -> Result<(), String> {
+	let directive = DEFAULT_DIRECTIVES
+		.get_or_init(|| Mutex::new(Vec::new()))
+		.lock()
+		.clone();
+
+	*CURRENT_DIRECTIVES
+		.get_or_init(|| Mutex::new(Vec::new()))
+		.lock() = directive;
+	reload_filter()
+}
+
+/// Initialize FILTER_RELOAD_HANDLE, only possible once
+pub(crate) fn set_reload_handle(handle: Handle<EnvFilter, SCSubscriber>) {
+	let _ = FILTER_RELOAD_HANDLE.set(handle);
+}
+
+// The layered Subscriber as built up in `init_logger()`.
```

Comment:
It's `GlobalLogger::init()` now yes?

Reply:
:scream:

Reply:
I wonder how useful it'd be to leverage rustdocs more consistently than we usually do, e.g. given `bla bla [`init_logger`] bla bla` rustdocs after rust 1.48 should link to `init_logger` (if it's in scope in this file). I'm not sure if it warns whenever the link goes out of date though. It definitely could and that'd be neat.

Reply:
I believe it doesn't at the moment but probably it will happen in the future because I heard already other people asking for this